include/sundials/sundials_context.hpp

```diff
@@ -21,47 +21,37 @@
 #define _SUNDIALS_CONTEXT_HPP
 
 #include <memory>
+#include <type_traits>
+
+#include <sundials/sundials_classview.hpp>
 #include <sundials/sundials_context.h>
 #include <sundials/sundials_convertibleto.hpp>
-
-#include "sundials/sundials_types.h"
+#include <sundials/sundials_types.h>
 
 namespace sundials {
 
-class Context : public sundials::ConvertibleTo<SUNContext>
+struct SUNContextDeleter
 {
-public:
-  explicit Context(SUNComm comm = SUN_COMM_NULL)
-  {
-    sunctx_ = std::make_unique<SUNContext>();
-    SUNContext_Create(comm, sunctx_.get());
-  }
-
-  /* disallow copy, but allow move construction */
-  Context(const Context&) = delete;
-  Context(Context&&)      = default;
-
-  /* disallow copy, but allow move operators */
-  Context& operator=(const Context&) = delete;
-  Context& operator=(Context&&)      = default;
-
-  SUNContext Convert() override { return *sunctx_.get(); }
-
-  SUNContext Convert() const override { return *sunctx_.get(); }
-
-  operator SUNContext() override { return *sunctx_.get(); }
+  void operator()(SUNContext sunctx) { SUNContext_Free(&sunctx); }
+};
 
-  operator SUNContext() const override { return *sunctx_.get(); }
+class SUNContextView
+  : public sundials::experimental::ClassView<SUNContext, SUNContextDeleter>
+{
+public:
+  using ClassView<SUNContext, SUNContextDeleter>::ClassView;
 
-  ~Context()
+  SUNContextView(SUNComm comm = SUN_COMM_NULL)
   {
-    if (sunctx_) { SUNContext_Free(sunctx_.get()); }
+    SUNContext sunctx = nullptr;
+    SUNContext_Create(comm, &sunctx);
+    object_.reset(sunctx, SUNContextDeleter{});
   }
```

Comment:
Untested, but something like this should prevent double initialization of the `shared_ptr`.
```suggestion
  SUNContextView(SUNComm comm = SUN_COMM_NULL) : SUNContextView(create(comm)) {}

private:
  static SUNContext create(SUNComm comm) noexcept {
    SUNContext sunctx = nullptr;
    SUNContext_Create(comm, &sunctx);
    return sunctx;
  }
```